### PR TITLE
Resolve patterns in inputs

### PR DESCRIPTION
**Pull request by paulben**

### CODE REVIEW

File: src/main/java/application/rest/v1/KAppNavEndpoint.java [hunk 1]
```diff
@@ -314,6 +320,196 @@ public static String getComponentNamespace(JsonObject component) {
...
+     * @param map        action ConfigMap
+     * @param client     Kube API client
+     * @param registry   ComponentInfoRegistry
+     * @param kind       target resource kind
+     * @param apiVersion target resource apiVersion
+     * @param name       target resource name
+     * @param namespace  target resource namespace
+     */
+    public static void resolveInputPatterns(JsonObject map, final ApiClient client, ComponentInfoRegistry registry, String kind, String apiVersion, String name, String namespace) {
+        final String methodName = "resolveInputsFuncs";
```
Comment (mrglavas) [inline src/main/java/application/rest/v1/KAppNavEndpoint.java]: Method name passed to the logger should match the actual method name. resolveInputsFuncs --> resolveInputPatterns.

File: src/main/java/application/rest/v1/KAppNavEndpoint.java [hunk 2]
```diff
@@ -314,6 +320,196 @@ public static String getComponentNamespace(JsonObject component) {
...
+        if (!kind.equals("Deployment") && !kind.equals("Pod")) {
+            if (Logger.isExitEnabled()) {
+                Logger.log(className, methodName, Logger.LogType.EXIT, "Only kind Deployment and Pod supported");
+            }
+            return;
+        }
+        try {
+            JsonObject resource = getResource(client, registry, name, kind, apiVersion, namespace);
+            boolean found = false;
+            try {
```
Comment (mrglavas) [inline src/main/java/application/rest/v1/KAppNavEndpoint.java:352]: With a simple null check and assignment there's no way an exception could be thrown here. Did you mean for the call to getResource() to be within the inner try/catch?

File: src/main/java/application/rest/v1/KAppNavEndpoint.java [hunk 3]
```diff
@@ -314,6 +320,196 @@ public static String getComponentNamespace(JsonObject component) {
...
+                                                // {"pods":["demo-app-56dcb8d858-sf2md"]}
+                                                if (podsArray != null) {
+                                                    for (JsonElement pod : podsArray) {
+                                                        valuesArray.add(pod);
+                                                        if (Logger.isDebugEnabled()) {
+                                                            Logger.log(className, methodName, Logger.LogType.DEBUG, "Added pod: " + pod.getAsString());
+                                                        }
+                                                    } 
+                                                }
+                                                rv = context.resolve("${resource.$.metadata.name}");
```
Comment (mrglavas) [inline src/main/java/application/rest/v1/KAppNavEndpoint.java]: Is this just for debugging purposes? The value isn't used anywhere else. Should probably put this resolve() call under the guard if it was only intended to be used by the logger.

File: src/main/java/application/rest/v1/KAppNavEndpoint.java [hunk 4]
```diff
@@ -314,6 +320,196 @@ public static String getComponentNamespace(JsonObject component) {
...
+                                    JsonArray valuesArray = field.getAsJsonArray("values");
+                                    if (valuesArray != null) {
+                                        if (Logger.isDebugEnabled()) {
+                                            Logger.log(className, methodName, Logger.LogType.DEBUG, "values array found");
+                                        }
+                                        try {
+                                            if (valuesArray.size() == 1 && valuesArray.get(0).getAsString().equals("${func.podlist()}")) {
+                                                ResolvedValue rv = context.resolve(valuesArray.get(0).getAsString());
+                                                if (Logger.isDebugEnabled()) {
+                                                    Logger.log(className, methodName, Logger.LogType.DEBUG, "values value: " + valuesArray.get(0).getAsString());
```
Comment (mrglavas) [inline src/main/java/application/rest/v1/KAppNavEndpoint.java:414]: Not sure if this was your intention, but this debug statement is essentially a static message. valuesArray.get(0).getAsString() always equals "${func.podlist()}".

**Review by mrglavas — CHANGES_REQUESTED** [review]
Did a high level review with Paul earlier today. Here are some comments from a more detailed review I just completed.

File: src/main/java/application/rest/v1/KAppNavEndpoint.java (hunk 1 above)
Comment (paulben) [reply src/main/java/application/rest/v1/KAppNavEndpoint.java]: Good catch. Fixed.

File: src/main/java/application/rest/v1/KAppNavEndpoint.java (hunk 2 above)
Comment (paulben) [reply src/main/java/application/rest/v1/KAppNavEndpoint.java:352]: Yes, that is what I intended. Fixed.

File: src/main/java/application/rest/v1/KAppNavEndpoint.java (hunk 3 above)
Comment (paulben) [reply src/main/java/application/rest/v1/KAppNavEndpoint.java]: That was code that I copied elsewhere and should have deleted here. Fixed - deleted.

File: src/main/java/application/rest/v1/KAppNavEndpoint.java (hunk 4 above)
Comment (paulben) [reply src/main/java/application/rest/v1/KAppNavEndpoint.java:414]: I have the log statement use the function for the potential case where the code handles all patterns.

**Review by mrglavas — APPROVED** [review]
Looks good now. Thanks.

**Pull request merged**